### ISSUE TITLE
When min major R2R version moves forward, consider some cleanups

### ISSUE THREAD

**Issue by VSadov** [dotnet/runtime#98871]
GCInfo version v3 is backward-compatible with v2. That is intentional to not require moving forward the R2R min major version.

However, if the min major version is incremented for other reasons, consider the following (not backward-compatible) changes -

* Safe points are historically encoded with offset -1. It is no longer necessary and not doing that could simplify some code.
* GCInfo v3 is sufficient that hijacking could be done without knowing whether a method is object/ref returning. 
Once v2 is no longer in use, both GCInfo and hijacking logic could be simplified.
* Actually anything that is keyed off !`AreSafePointsInterruptible` could be removed, except when needed to support x86.
(i.e. anything for RISC architectures can stop supporting noninterruptible call sites)
* Also consider removing `DOTNET_InterruptibleCallSites` knob at that time

These suggestions are not by themselves enough to bump the min major version, but the version moves forward for other reasons, please consider doing these as well.

**Comment by EgorBo** [comment]
Some helpers can be removed too, https://github.com/dotnet/runtime/pull/98858#issuecomment-1962073034

**Comment by VSadov** [comment]
Right. Let's keep this as a holder for cleanups, conditional on revving min version. 
When that happens we can split this into multiple actionable workitems for the milestone when that happens.

**Comment by filipnavara** [comment]
Likewise there are some ARM helpers that are now obsolete. We marked them with a comment in the appropriate source code.

**Comment by MichalPetryka** [comment]
The delegate method pointers (which are being swapped in #99200 to avoid R2R bump) can be swapped back on the next bump for some minor perf improvements.

**Comment by davidwrighton** [comment]
I now have a PR #100310 which represents the cleanups.

I've talked with @VSadov, and unfortunately, it looks like the V3 GCInfo isn't happening due to an issue with our codegen in RyuJit that won't get fixed this release. 

@MichalPetryka I know you've done some work on the math helpers, I can see that we only want to remove the floating point rounding helpers instead of the 4 helpers you initially proposed as making ready for removal.

Since #99200 is not yet in, I haven't put together the changes to re-order the fields.

**Comment by VSadov** [comment]
> I've talked with @VSadov, and unfortunately, it looks like the V3 GCInfo isn't happening due to an issue with our codegen in RyuJit that won't get fixed this release.

The issue might be fixable. I am still looking at how it may be dealt with. 

There are some obvious solutions like placing a NOP at the end of every basic block that ends on a method call, but I'd prefer a solution that is less intrusive (which I did not figure yet).

**Comment by VSadov** [comment]
With #110845 this can be considered completed.
